[ES|QL] Mark MATCH_PHRASE second argument as constantOnly

bartoval opened the pull request:
## Summary

I use this PR https://github.com/elastic/kibana/issues/239209 to implement some improvements to the suggestion system


https://github.com/user-attachments/assets/de074b2f-acd8-49a8-aa44-ace4d0c02d1f

#### 1. Added `constantOnly: true` to MATCH_PHRASE second parameter

**File:** `generated/scalar_functions.ts`

This flag tells the autocomplete system: "only literal values are accepted here".

**Effect:** When the user is at `MATCH_PHRASE(field, |)`,  not fields are suggested.

---

#### 2. Excluded all FTS functions from EVAL

**File:** `eval/autocomplete.ts`

FTS functions (match, match_phrase, kql, qstr, multi_match, match_operator) don't make sense in EVAL alone

**Exception:** They're allowed inside `SCORE()`, which is the only valid EVAL context.

**Why ?**  When you run a query with these functions in FTS functions, ES throws an error

---

#### 3. Added nested function support

**File:** `in_function.ts`

When a cursor is inside a nested function, (eg. `OUTER(INNER(field, |))`), the system needs to recognize that suggestions should come from the inner function and not the outer one

I added a check that verifies if the cursor position is within the bounds of the last argument. If so, it recursively enters the nested function.

---

#### 4. Added placeholder for constantOnly parameters

**Files:** `empty_expression.ts`, `complete_items.ts`

When a parameter only accepts literals, I show a hint like "Insert value placeholder" to guide the user. It can be a number , a string or Object

**Why?** A field that only accepts constants always returned empty suggestions or formulas in rare cases. It seemed confusing to me not to give suggestions.

---

#### 5. Refactoring: separating tracking from filtering in expressionsForSuggestions

**Files:** `in_function.ts`, `empty_expression.ts`, `types.ts`, `suggestion_builder.ts`

**Before:** `FunctionParameterContext` contained a `functionsToIgnore: string[]` that accumulated parent function names to prevent recursion (e.g., don't suggest ABS inside ABS). This mixed recursion-prevention data with parameter information.

**After:** I separated into two distinct concepts in `ExpressionContextOptions`:

- **`parentFunctionNames`**: tracks the chain of functions we're nested inside (for recursion prevention)
- **`functionsToIgnore`**: command-level configuration specifying which functions to hide and where they're allowed as exceptions (e.g., EVAL hides FTS functions, except inside SCORE)

`SuggestionBuilder.resolveIgnoredFunctions()` combines both sources to determine what to exclude.

**Benefit:** Separation of concerns. Recursion prevention is separate from command-specific filtering rules.

---

#### 6. STATS refactoring: filtering context separation

**File:** `stats/autocomplete.ts`

**Before:** STATS mixed its own filtering context (grouping functions, conflicting aggregate functions) inside `FunctionParameterContext`:

STATS passes separately:
- `functionParameterContext`: info about current function parameters
- `functionsToIgnore`: functions to exclude (grouping, conflicting aggregates)

**Benefit:** `FunctionParameterContext` stays focused on parameters and we can generalize `functionToIgnore logic.

---

#### 7. Extended complete_items for typed placeholders

**File:** `complete_items.ts`

I introduced type-specific placeholders: strings (`"${0:value}"`), numbers (`${0:0}`), config objects (`{ $0 }`). The function `findConstantPlaceholderType(types)` selects the appropriate one based on accepted ES|QL types.

**Why?** Type-aware placeholders prevent confusion and make the system extensible.

---


#### 9. Readability improvements

**Files:** `signature_analyzer.ts`, `suggestion_engine.ts`, `map_expression.ts`, `position.ts`

`map_expression.ts` only added the logic needed to handle the new `complete item` and to reuse some functionality from `SuggestForExpression`. The rest is just a reorganization of the functions.

bartoval committed:
match phrase second argument must be constant only

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-entity-store-performance-from-pr --> Click to trigger **kibana-entity-store-performance-from-pr** for this PR!

bartoval commented on src/platform/packages/shared/kbn-esql-ast/src/__tests__/commands/autocomplete.ts:
```diff
@@ -44,9 +44,19 @@ import type {
 import { correctQuerySyntax, findAstPosition } from '../../commands/definitions/utils/ast';
 
+// Functions to ignore per location (e.g., match_phrase is only shown inside SCORE, not in EVAL)
+const HIDDEN_FTS_FUNCTIONS: FullTextSearchFunctionName[] = ['match_phrase'];
```
It's an adaptation to the test expectations (many of them failed)

bartoval commented on src/platform/packages/shared/kbn-esql-ast/src/commands/registry/eval/autocomplete.ts:
```diff
@@ -20,6 +21,13 @@ import {
 import type { ICommandCallbacks } from '../types';
 import { Location, type ICommandContext, type ISuggestionItem } from '../types';
 
+const FUNCTIONS_TO_IGNORE = {
```
We pass a configuration at the command level that lets us decide what to show and what to hide, and allows us to skip this role for certain inner cases.

bartoval committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

stratoula committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

stratoula commented on src/platform/packages/shared/kbn-esql-language/src/commands/definitions/utils/autocomplete/expressions/position.ts:
```diff
@@ -90,5 +90,11 @@ export function getPosition(
 
+  // Map expressions (e.g., {"key": "value"}) are complete expressions
+  // Like literals and columns, a closed map is a fully-formed value
+  if (expressionRoot.type === 'map') {
```
Why don't you combine it with the above statement?

stratoula commented on src/platform/packages/shared/kbn-esql-language/src/commands/definitions/utils/autocomplete/map_expression.ts:
Why are we doing these map expression changes?

stratoula commented on src/platform/packages/shared/kbn-esql-language/src/commands/definitions/utils/autocomplete/map_expression.ts:
```diff
@@ -42,83 +50,126 @@ export type MapParameters = Record<string, MapParameterValues>;
-    while ((match = regex.exec(innerText)) !== null) {
-      usedParams.add(match[1]);
-    }
+    const usedParams = new Set([...innerText.matchAll(/"([^"]+)"\s*:/g)].map(([, name]) => name));
```
Can you add it in one constant to be more readable? (also to have consistency) I know it was like that but can we also add it in a constant as the DOUBLE_QUOTE one?

stratoula reviewed (commented):
Valerio I can see that we did various changes that I can't follow. Can you update the description to be more clear what we changed and why?

This will make the review a bit easier. I can confirm that the autocomplete of the match and match_phrase has been improved but what else did we fix?

bartoval commented:
> Valerio I can see that we did various changes that I can't follow. Can you update the description to be more clear what we changed and why?
> 
> This will make the review a bit easier. I can confirm that the autocomplete of the match and match_phrase has been improved but what else did we fix?

I've updated the description. I used this task to do some experimentation. I can also just stick to the task context and solve it with a few changes.

Quickly:
- The issue mentions match_phrase, but I don’t believe match or similar commands have the same restrictions.

- The autocomplete changes apply to all functions, not only match and match_phrase.
For example, in bucket, the autocomplete now suggests a placeholder for an integer, which is one of the supported mixed signature options.

- Changes related to maps are minimal. Most of the logic has been extracted into utility functions, and I adapted it to support maps within suggestForExpression.

bartoval committed:
update feedbacks

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-entity-store-performance-from-pr --> Click to trigger **kibana-entity-store-performance-from-pr** for this PR!

bartoval committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

bartoval committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

bartoval committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

bartoval committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

stratoula commented on src/platform/packages/shared/kbn-esql-language/src/commands/definitions/utils/autocomplete/expressions/signature_analyzer.ts:
```diff
@@ -443,4 +446,39 @@ export class SignatureAnalyzer {
+
+      const values = rawValues
+        .split(',')
+        .map((val) => val.trim().replace(/^'|'$/g, ''))
```
Move the regex to a constant to be more understandable what it does

stratoula commented on src/platform/packages/shared/kbn-esql-language/src/commands/registry/eval/autocomplete.ts:
```diff
@@ -20,6 +21,13 @@ import {
+export const FUNCTIONS_TO_IGNORE = {
+  names: [...FULL_TEXT_SEARCH_FUNCTIONS],
+  allowedInsideFunctions: Object.fromEntries(
+    FULL_TEXT_SEARCH_FUNCTIONS.map((fn) => [fn, ['scsore']])
```
Is scsore a typo?

stratoula committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

stratoula commented on src/platform/packages/shared/kbn-esql-language/src/__tests__/commands/autocomplete.ts:
Shouldn't the match phrase been suggested here? 

<img width="1218" height="357" alt="image" src="https://github.com/user-attachments/assets/1be01b2a-9c55-48b6-97d6-2a47cd6b4912" />

bartoval replied:
it was due to the typo error. now it works

stratoula reviewed (commented):
Thanks for the extended description Val. It was helpful. Next time let's try to have separate PRs though, it would make the review faster I think.

It looks really good and I love the enhancements. Added some comments, nothing crucial

bartoval committed:
update comments

bartoval committed:
Merge branch 'main' into Match_phrase_second_argument_constant_only

stratoula reviewed (approved):
I didnt test again, LGTM! Thanks Val
